### PR TITLE
Make `IServiceScopeFactory` a singleton, child scopes are flat

### LINKED ISSUE #83: IServiceScopeFactory should be a singleton; service scopes should not be nested
**Describe the bug**
If [AutofacServiceScopeFactory](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/master/src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs#L97) is transient, that will not compatible with Ms DI.

**To reproduce**
``` C#
if (provider.GetService<IServiceScopeFactory>() == provider.GetService<IServiceScopeFactory>())
{
    //Ms DI.
}
else
{
    //Autofac
}
```

### PR DESCRIPTION
**Pull request by tillig**
Fix #83.

**BREAKING CHANGE**: Makes `IServiceScopeFactory` a singleton and switches child lifetime scopes to be flat (all come from the root container) rather than hierarchical based on https://github.com/dotnet/runtime/issues/67391. Users wanting to create nested/hierarchical scopes will need to inject `ILifetimeScope` instead of `IServiceScopeFactory` to create scopes.

PR includes update of semantic version to 8.0.0 to account for the breaking change. Also now testing Microsoft.Extensions.DependencyInjection both through the manual registrations (e.g., calling `builder.Populate()`) and through the service provider factory implementations to ensure all mechanisms are consistently compatible.

The `IServiceScopeFactory` singleton test and flat lifetime scope test added here to our `AssumedBehaviorTests` are also submitted [as a PR to the dotnet/runtime repo](https://github.com/dotnet/runtime/pull/68517).

This PR also brings the build/analyzers up to date with core Autofac including:

- Update .editorconfig to align with core Autofac.
- Update analyzer rulesets to align with core Autofac (including file headers / stylecop.json).
- Publish symbols as .snupkg.
- Ability to run benchmarks via .exe or Xunit.

### CODE REVIEW

File: test/Autofac.Extensions.DependencyInjection.Test/Specification/AssumedBehaviorTests.cs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        }
+
+        Assert.True(tracker.AsyncDisposed);
+        Assert.False(tracker.SyncDisposed);
+    }
+
+#endif
+
+    [Fact]
+    public void ServiceScopeFactoryIsSingleton()
```
Comment (tillig) [inline test/Autofac.Extensions.DependencyInjection.Test/Specification/AssumedBehaviorTests.cs:145]: Test for `IServiceScopeFactory` being singleton.

File: test/Autofac.Extensions.DependencyInjection.Test/Specification/AssumedBehaviorTests.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        var rootFactory2 = rootProvider.GetRequiredService<IServiceScopeFactory>();
+        Assert.Same(rootFactory1, rootFactory2);
+
+        var childScope = rootFactory1.CreateScope();
+        var childFactory = childScope.ServiceProvider.GetRequiredService<IServiceScopeFactory>();
+        Assert.Same(rootFactory1, childFactory);
+    }
+
+    [Fact]
+    public void ServiceScopesAreFlat()
```
Comment (tillig) [inline test/Autofac.Extensions.DependencyInjection.Test/Specification/AssumedBehaviorTests.cs:161]: Test for scope hierarchy being flat.

**Comment by tillig** [comment]
I'm not sure how coverage went _down_ by adding more tests and, like, adding braces around single-line statements but that's weird. I'm planning a follow-up here before releasing this to:

- Figure out why our tests don't pass the `IServiceProviderIsService` compatibility tests in the dotnet/runtime repo.
- Switch to file-scoped namespaces.
- Switch to `ArgumentNullException.ThrowIfNull()`.
- Test this on .NET 7 preview and verify there's nothing else we need to add support for.

I don't think coverage at the moment should block this, I'll fix it up after the fact.

**Review by alistairjevans — APPROVED** [review]
As you say, not actually a super-complicated change (for us, anyway).

**Comment by alsami** [comment]
Will ths be released any time soon? I could then integrate it in the multitenant package for ASP.NET Core. My guess is that nothing obvious will break but better safe than sorry :)

**Comment by tillig** [comment]
As I noted earlier, I need to test against .NET 7 preview and also figure out why we're not passing the IServiceProviderIsService tests over in the main dotnet repo. We can release after that. But the day job has me a little swamped so "soon" isn't going to be the next day or two. But... Should be not too long. I hope? You could consume a build from MyGet to get a head start.

**Comment by tillig** [comment]
8.0.0 went out just now. Have at it!

**Comment by alsami** [comment]
I am on a short holiday until Sunday but after I will take on the task

**Pull request merged**